File: neuron_db/router.py
"""NeuronRouter -- chain many neurons into one memory.

A single neuron recalls best when it holds distinct, not-too-many facts (see
BENCHMARKS.md: cue collisions grow with size). The router sidesteps that ceiling by
spreading facts across many small neuron *shards* and fanning a query out across them,
then returning the single best-matching value.

This is the answer to "can we chain neurons instead of stuffing everything into a giant
context?" Yes: the model never sees the whole store -- it asks a question and gets one
value back, no matter how many shards there are.

    r = NeuronRouter(per_shard=128)
    for fact in many_facts: r.observe(fact)     # auto-spills into new shards
    r.recall("what is the north gate code?")    # fan-out, best value -> one answer
"""
from __future__ import annotations
from typing import Optional
from .neuron import Neuron


class NeuronRouter:
    def __init__(self, per_shard: int = 128):
        self.per_shard = per_shard
        self.shards: list[Neuron] = [Neuron(max_facts=10 ** 9)]  # shards don't self-truncate; router controls size

    # --- write: fill the current shard, spill to a new one when full ---
    def observe(self, text: str) -> int:
        if self.shards[-1].fact_count >= self.per_shard:
            self.shards.append(Neuron(max_facts=10 ** 9))
        return self.shards[-1].observe(text)

    # --- read: fan out across shards, keep the strongest hit ---
    def recall(self, query: str) -> Optional[dict]:
        best = None; bk = (-1, -1, -1.0)  # (exact, overlap, coverage)
        for sh in self.shards:
            hit = sh.recall(query)
            if hit is None:
                continue
            sc = (hit.get("exact", 0), hit["overlap"], hit["coverage"])
            if sc > bk:
                bk = sc; best = hit
        return best

    def get(self, query: str) -> Optional[str]:
        hit = self.recall(query)
        return hit["value"] if hit else None

    @property
    def fact_count(self) -> int:
        return sum(s.fact_count for s in self.shards)

    @property
    def shard_count(self) -> int:
        return len(self.shards)

    def dump(self) -> list:
        return [s.dump() for s in self.shards]

    @classmethod
    def load(cls, blobs: list, per_shard: int = 128) -> "NeuronRouter":
        r = cls(per_shard)
        r.shards = [Neuron.load(b, max_facts=10 ** 9) for b in blobs] or [Neuron(max_facts=10 ** 9)]
        return r

File: rust/neuron-core/src/router.rs
//! NeuronRouter: chain many small neuron shards into one memory and fan a query out across
//! them. A single neuron's recall degrades as it grows (common cues collide); the router
//! keeps each shard small and picks the single best-matching value across all shards.
//! Faithful std-only port of neuron_db/router.py.

use crate::{Neuron, Recall};

const BIG: usize = usize::MAX / 2; // shards never self-truncate; the router controls size

pub struct NeuronRouter {
    pub per_shard: usize,
    pub shards: Vec<Neuron>,
}

impl NeuronRouter {
    pub fn new(per_shard: usize) -> Self {
        NeuronRouter { per_shard, shards: vec![Neuron::new(BIG)] }
    }

    /// Fill the current shard; spill into a fresh one when it reaches `per_shard`.
    pub fn observe(&mut self, text: &str) -> usize {
        if self.shards.last().unwrap().fact_count() >= self.per_shard {
            self.shards.push(Neuron::new(BIG));
        }
        self.shards.last_mut().unwrap().observe(text)
    }

    /// Fan out across shards, keep the strongest hit by (overlap, coverage).
    pub fn recall(&mut self, query: &str) -> Option<Recall> {
        let mut best: Option<Recall> = None;
        let mut bk = (-1i64, -1i64, -1f64);
        for sh in self.shards.iter_mut() {
            if let Some(h) = sh.recall(query) {
                let sc = (h.exact as i64, h.overlap as i64, h.coverage);
                let better = sc.0 > bk.0 || (sc.0 == bk.0 && sc.1 > bk.1)
                    || (sc.0 == bk.0 && sc.1 == bk.1 && sc.2 > bk.2);
                if better { bk = sc; best = Some(h); }
            }
        }
        best
    }
    pub fn get(&mut self, query: &str) -> Option<String> { self.recall(query).map(|h| h.value) }
    pub fn fact_count(&self) -> usize { self.shards.iter().map(|s| s.fact_count()).sum() }
    pub fn shard_count(&self) -> usize { self.shards.len() }
    pub fn dump(&self) -> Vec<String> { self.shards.iter().map(|s| s.dump()).collect() }
    pub fn load(blobs: &[String], per_shard: usize) -> Self {
        let shards: Vec<Neuron> = blobs.iter().map(|b| Neuron::load(b, BIG)).collect();
        NeuronRouter { per_shard, shards: if shards.is_empty() { vec![Neuron::new(BIG)] } else { shards } }
    }
}
